jax/interpreters/mlir.py

```diff
@@ -926,13 +927,19 @@ def write(v: core.Var, node: Sequence[ir.Value]):
             f"found for platform {ctx.platform}")
 
       eqn_ctx = (ctx.replace(name_stack=source_info.name_stack) if
-          config.jax_experimental_name_stack else ctx)
+                 config.jax_experimental_name_stack else ctx)
       effects = [eff for eff in eqn.effects if eff in core.ordered_effects]
       tokens_in = tokens.subset(effects)
+      avals_in = map(aval, eqn.invars)
       rule_ctx = LoweringRuleContext(
-          module_context=eqn_ctx, primitive=eqn.primitive,
-          avals_in=map(aval, eqn.invars), avals_out=map(aval, eqn.outvars),
-          tokens_in=tokens_in, tokens_out=None)
+          module_context=eqn_ctx, primitive=eqn.primitive, avals_in=avals_in,
+          avals_out=map(aval, eqn.outvars), tokens_in=tokens_in,
+          tokens_out=None)
+      if config.jax_dynamic_shapes:
+        axis_size_env = {d: read(d)[0] for a in avals_in
+                         if type(a) is core.DShapedArray for d in a.shape
+                         if type(d) is core.Var}
+        rule_ctx = rule_ctx.replace(axis_size_env=axis_size_env)
```

Comment:
As mentioned above (in the context of the changes to `mhlo_broadcast`), we extend the `rule_ctx` to include an `axis_size_env: Dict[core.Var, ir.Value]` with keys corresponding to the `Var` instances that occur in the shapes of the `avals_in`. (We might need to add those in `avals_out` also...)